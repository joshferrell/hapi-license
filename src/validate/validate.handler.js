import boom from 'boom';

/* eslint-disable import/prefer-default-export */

export const createValidateLicense = (LicenseModel, validateLicense) => (request, reply) => {
    const { license, computerId } = request.payload;

    return LicenseModel
        .findOne({
            where: { license },
            raw: true
        })
        .then(model => (
            model ?
                validateLicense(license, computerId) :
                Promise.reject(new Error('license does not exist'))
        ))
        .then(() => reply({ success: true }))
        .catch((err) => {
            switch (err) {
            case 'computer id does not match license':
                return reply(boom.unauthorized());
            case 'license does not exist':
                return reply(boom.unauthorized());
            default:
                request.log('error', {
                    err,
                    msg: 'unable to validate license'
                });
                return reply(boom.badImplementation());
            }
        });
};
