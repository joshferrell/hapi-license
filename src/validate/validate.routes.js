import boom from 'boom';
import joi from 'joi';
import { format } from '../utility';

const createValidateRoutes = () => {
    const notImplemented = (request, reply) => reply(boom.notImplemented());

    return [
        {
            method: 'POST',
            path: '/validate',
            handler: notImplemented,
            config: {
                auth: false,
                validate: {
                    payload: {
                        license: joi.string().required()
                            .description('A token to validate'),
                        computerId: joi.string().required()
                            .description('Current computers identifier')
                    }
                },
                tags: ['api', 'License Validation'],
                description: 'Validate that a license is valid by id',
                notes: [
                    'Validates that a license is valid or active'
                ],
                plugins: {
                    'hapi-swagger': {
                        responses: {
                            ...format.success,
                            ...format.badRequest,
                            ...format.unauthorized,
                            ...format.internalError,
                            ...format.notImplemented
                        }
                    }
                }
            }
        }
    ];
};

export default createValidateRoutes;
