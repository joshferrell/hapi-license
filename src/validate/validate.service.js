import { promisify } from 'util';
import jwt from 'jsonwebtoken';

const verify = promisify(jwt.verify);

/* eslint-disable import/prefer-default-export */

export const createLicenseValidator = secret => (token, computerId) =>
    verify(token, secret)
        .then(decodedToken => (
            decodedToken.computer.computerId === computerId ?
                Promise.resolve(decodedToken) :
                Promise.reject(new Error('computer id does not match license'))
        ));
