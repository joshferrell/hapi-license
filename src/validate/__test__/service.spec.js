import jwt from 'jsonwebtoken';
import { promisify } from 'util';
import { createLicenseValidator } from '../validate.service';

const sign = promisify(jwt.sign);

const signToken = body => sign(body, 'shhh', { algorithm: 'HS256' });

describe('validation services', () => {
    describe('license validator', () => {
        let tokenBody;
        let licenseValidate;
        beforeEach(() => {
            tokenBody = {
                accountId: '1234',
                computer: {
                    computerId: 'computer-id'
                },
                productId: 'product something'
            };

            licenseValidate = createLicenseValidator('shhh');
        });

        it('should verify a valid token', async () => {
            const token = await signToken(tokenBody);
            const decoded = await licenseValidate(token, 'computer-id');

            expect(decoded.accountId).toEqual('1234');
            expect(decoded.computer).toEqual({
                computerId: 'computer-id'
            });
            expect(decoded.productId).toEqual('product something');
        });

        it('should reject if the computer id is not in the token', async () => {
            tokenBody.computer.computerId = 'bad to the bone';
            const token = await signToken(tokenBody);
            return expect(licenseValidate(token, 'computer-id')).rejects.toMatchSnapshot();
        });
    });
});
