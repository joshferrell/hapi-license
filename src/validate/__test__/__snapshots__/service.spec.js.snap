// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validation services license validator should reject if the computer id is not in the token 1`] = `[Error: computer id does not match license]`;
